src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -79,41 +82,42 @@ export default class Shoha extends Module {
 			},
 			this.onSpend,
 		)
-		this.addEventHook([
-			'applybuff', 'applybuffstack'],
-			{
-				to: 'player',
-				abilityId: STATUSES.MEDITATION.id,
-			},
-			this.onMeditateGain)
+		this.addHook('applybuff', {to: 'player', abilityId: STATUSES.MEDITATE.id}, this.onApplyMeditate)
+		this.addHook('removebuff', {to: 'player', abilityId: STATUSES.MEDITATE.id}, this.onRemoveMeditate)
 
 		this.addHook('death', {to: 'player'}, this.onDeath)
 		this.addHook('complete', this.onComplete)
 	}
 
 	private onGenerate(event: CastEvent) {
-		if (this.stacks === MAX_STACKS) {
 		const abilityId = event.ability.guid
 		const generatedStacks = GENERATORS[abilityId]
 
-		this.addWaste(generatedStacks, abilityId)
+		this.addStacks(generatedStacks, abilityId)
 		}
+
+		// back to time guessing we a go~
+	private onApplyMeditate(event: BuffEvent) {
+			this.meditateStart = event.timestamp
 	}
 
-		// the previous comment was a lie: This is for ALL meditation gains
-	private onMeditateGain() {
-			this.stacks++
-			this.totalGeneratedStacks++
+	private onRemoveMeditate(event: BuffEvent) {
+		const diff = event.timestamp - this.meditateStart
+
+		const ticks = Math.min(Math.floor(diff / MEDITATE_TICK_FREQUENCY), MAX_MEDITATE_TICKS)
+
+		const generatedStacks = ticks
 
-			// just a safety net
-			if (this.stacks > MAX_STACKS) {
-				this.stacks = MAX_STACKS
-			}
+		this.stacks += generatedStacks
+		this.totalGeneratedStacks += generatedStacks
```

Comment:
Are you counting total generated stacks including overcapping? If not, might wanna move this to only happen when `this.stacks <= MAX_STACKS`